### PR TITLE
Only Schedule Retries before Agent Shutdown

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35554 Only Schedule Retries before Agent Shutdown**

We attach a callback to our RPC send attempts that schedule a retry
upon failure. This PR only schedules the retry if the agent is running.

Fixes: https://github.com/pytorch/pytorch/issues/35815

Differential Revision: [D20612615](https://our.internmc.facebook.com/intern/diff/D20612615/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20612615/)!

### CODE REVIEW

**Comment by rohan-varma** [comment]
In the retry thread, should we also add a try-catch around `send`, so that RPCs that are already retrying won't throw and result in a `std::terminate`?

**Comment by osalpekar** [comment]
> In the retry thread, should we also add a try-catch around `send`, so that RPCs that are already retrying won't throw and result in a `std::terminate`?

Yeah, that's in a separate PR: https://github.com/pytorch/pytorch/pull/35263

File: torch/csrc/distributed/rpc/rpc_agent.cpp [hunk 1]
```diff
@@ -156,7 +156,8 @@ void RpcAgent::rpcRetryCallback(
     // just the retry count.
     LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_ + 1)
               << " failed";
-    if (earliestRpc->retryCount_ < earliestRpc->options_.maxRetries) {
+    if (earliestRpc->retryCount_ < earliestRpc->options_.maxRetries &&
+        rpcAgentRunning_.load()) {
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rpc_agent.cpp]: as we don't hold lock here, this is best_effort check? Could it happen that `rpcAgentRunning_` toggle to false immediately after this check? Will that cause any issues?

Comment (osalpekar) [reply torch/csrc/distributed/rpc/rpc_agent.cpp]: Yes, this is a best-effort check, basically as a guard against scheduling retries if the agent has been shutdown. If `rpcAgentRunning_` is set to False immediately after this:
1. If the retry thread is processing retries, #35263 will ensure that performing sends when the agent is shutdown is handled gracefully.
1. If the retry thread is not processing retries (sleeping), then the retry function will simply return upon waking up and not process any further retries (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rpc_agent.cpp#L85).

**Review by mrshenli — APPROVED** [review]
Code LGTM!

Shall we do a rebase and wait for tests? #35283 should have fixed the error in `test_torchscript_functions_not_supported`. And some tests show merge conflicts.

**Pull request closed without merging**